src/validator.ts

```diff
@@ -2,49 +2,243 @@ import Ajv from 'ajv';
+const prepend = (prefix: string, message: string) => {
+  return `${prefix} ${message}`;
+};
+
+const schemaPathPropNameRegex = /properties\/(.+)\//;
+const schemaPathArrayPropNameRegex = /properties\/(.+)\/items/;
+const schemaPathGroupNumberRegex = /anyOf\/(\d+)\/(.+)/;
+const instancePathItemIndexRegex = /(\d+)$/;
+
+// If there are more than maxErrors errors in a group, they
+// will get summarized with an error count.
+const maxErrors = 3;
+
+const formatIndividualError = (e, formattedMessageList) => {
+  if (e.schemaPath.indexOf('properties') > -1) {
+    // property of an object
+    if (e.schemaPath.indexOf('items') > -1) {
+      // property is an array
+      const propNameMatch = schemaPathArrayPropNameRegex.exec(e.schemaPath);
+      const propName = propNameMatch ? propNameMatch[1] : 'unknown';
+      const itemIndexMatch = instancePathItemIndexRegex.exec(e.instancePath);
+      const itemIndex = itemIndexMatch ? itemIndexMatch[1] : 'unknown';
+      formattedMessageList.push(
+        `item at index ${itemIndex} of the '${propName}' array ${e.message}`
+      );
+    } else {
+      // property is not an array
+      const propNameMatch = schemaPathPropNameRegex.exec(e.schemaPath);
+      const propName = propNameMatch ? propNameMatch[1] : 'unknown';
+      formattedMessageList.push(`property '${propName}' ${e.message}`);
+    }
+  } else if (e.schemaPath.indexOf('items') > -1) {
+    // item in an array
+    const itemIndexMatch = instancePathItemIndexRegex.exec(e.instancePath);
+    const itemIndex = itemIndexMatch ? itemIndexMatch[1] : 'unknown';
+    formattedMessageList.push(
+      `item at index ${itemIndex} of the array ${e.message}`
+    );
+  } else if (e.instancePath === '') {
+    // parameter is something other than an object, e.g. string
+    formattedMessageList.push(`the argument ${e.message}`);
+  }
+};
+
+const missingPropertiesErrors = (
+  subject: string,
+  errors: Array<ErrorObject>,
+  messageParts: Array<string>
+) => {
+  const missingPropertyNames = errors
+    .filter((error) => error.keyword === 'required')
+    .map((error) => {
+      return error.params.missingProperty !== undefined
+        ? error.params.missingProperty
+        : 'unknown';
+    });
+  if (missingPropertyNames.length > 0) {
+    const missingMessage = prepend(
+      subject,
+      `must have required ${
+        missingPropertyNames.length > 1 ? 'properties' : 'property'
+      }: ${missingPropertyNames.join(', ')}.`
+    );
+    messageParts.push(missingMessage);
+  }
+};
+
+const typeErrors = (
+  subject: string,
+  errors: Array<ErrorObject>,
+  messageParts: Array<string>
 ) => {
+  const typeErrorsList: Array<string> = [];
+  let errorCount = 0;
+
+  for (let i = 0; i < errors.length; i++) {
+    const e = errors[i];
+
+    if (e.keyword === 'type') {
+      errorCount += 1;
+      if (errorCount > maxErrors) {
+        continue;
+      } else {
+        formatIndividualError(e, typeErrorsList);
+      }
```

Comment:
We don't want to `break` out of the loop because we need to continue counting them. But Anna's right that I could simplify this if-statement by inverting it.